### ISSUE TITLE
Dropout and DropConnect

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#373]
Ok, so dropout and dropconnect. During training we apply them (if we are using them), but during prediction we don't.
Some issues/observations:
(a) I don't see a distinction being made between training and prediction in practice, for example in MultiLayerNetwork.feedForward() and BaseLayer.activate(). For example, for drop connect: it's always applied in MultiLayerNetwork.feedForward().

(b) I'm not sure that dropout is even being applied in general, regardless of settings.
So for standard feed-forward networks, we have BaseLayer.applyDropOutIfNecessary(...). But as far as I can tell, the only place this is called is in RBM.sampleHiddenGivenVisible(), i.e., for RBM pretraining?
Also consider where NeuralNetConfiguration.getDropOut() is ever called: I can only find BaseLayer.applyDropOutIfNecessary(...) and in the LSTM layer.

(c) Also for drop connect, (see MultiLayerNetwork.applyDropConnectIfNecessary(INDArray input)) a binomial mask is created, but then this mask is applied to the current layer input. I was under the impression that weights are set to zero in DropConnect, not inputs/activation (the latter is dropout). See for example http://cs.nyu.edu/~wanli/dropc/

(d) [I'm not 100% sure on this; I'd suggest checking the literature.] When using dropout, weights should be scaled after training to ensure that the expected output/activation is the same during training (with dropout applied) vs. during prediction, right? (Not sure if rescaling should be applied with pretraining+dropout, backprop+dropout, or both?) I don't see this happening anywhere (though could have missed it).
(See also: comment on "inverted" dropout implementation for LSTM here: https://github.com/karpathy/neuraltalk/blob/master/imagernn/lstm_generator.py. This is what DL4J's LSTM layer is based on. This inverted version of dropout apparently avoids the need to rescale weights with dropout.)

(e) Not really an issue per-se, more of a design decision. The drop connect probability is currently fixed at 0.5 (see MultiLayerNetwork.applyDropConnectIfNecessary(INDArray input)), but this need not be the case in general. Do you want to allow people to use arbitrary drop connect probabilities?

**Comment by agibsonccc** [comment]
Fair points everywhere. Will look at those

Re: e. This is taken care of in the neuralnet configuration.

**Comment by agibsonccc** [comment]
@AlexDBlack I'll separate train/test when predict happens.

**Comment by agibsonccc** [comment]
https://github.com/deeplearning4j/deeplearning4j/pull/379

**Comment by AlexDBlack** [comment]
I was thinking that the best strategy is to use the "inverted dropout" approach throughout DL4J?
So with standard dropout, we either (a) have to scale weights a single time once training is done (which is bad: but how do we know training is finally completed?) OR (b) we have to scale them during every test time forward pass (also bad: for performance reasons + you need to know the training dropout configuration during test time).

I can't find a lot on the inverted method (I only came across it by accident) but see for example:
Li and/or Karpathy: "The undesirable property of the [standard dropout] scheme presented above is that we must scale the activations by p at test time. Since test-time performance is so critical, it is always preferable to use inverted dropout, which performs the scaling at train time, leaving the forward pass at test time untouched. Additionally, this has the appealing property that the prediction code can remain untouched when you decide to tweak where you apply dropout, or if at all."
http://cs231n.github.io/neural-networks-2/
http://vision.stanford.edu/teaching/cs231n/slides/lecture6.pdf

**Comment by agibsonccc** [comment]
Re: first point. The model needs to be updatable online too so training is never "done".

Re: second point. I was aware of a lot of that literature. That's why I only do it at train time. I'm going to go ahead and close then (drop connect also multiplies only the weights in the appropriate places)

**Comment by jpatanooga** [comment]
These series of discussions/debates around the training system are
important. We should write them out in a FAQ specifically for training.
They would also make a great appendix for our deep learning book as well.
